src/pagination.rs

```diff
@@ -0,0 +1,304 @@
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc
+        .front
+        .pagination
+        .as_ref()
+        .expect("Front should have pagination here.");
+    let all_posts: Vec<_> = posts_data.iter().collect();
+    match config.include {
+        Include::All => {
+            let all_posts = sort_posts(&all_posts, &config);
+            create_all_paginators(&all_posts, &doc, &config)
+        }
+        Include::None => {
+            unreachable!("PaginationConfigBuilder should have lead to a None for pagination.")
+        }
+    }
+}
+
+fn extract_value<'a>(a: &'a liquid::value::Value, key: &'static str) -> Option<&'a liquid::value::Scalar> {
+    a.get(&key.into()).and_then(|sort_key| sort_key.as_scalar())
+}
+
+// sort posts by multiple criteria
+fn sort_posts<'a>(
+    posts: &'a Vec<&liquid::value::Value>,
+    config: &PaginationConfig,
+) -> Vec<&'a liquid::value::Value> {
+    let order: fn(&liquid::value::Scalar, &liquid::value::Scalar) -> Ordering = match config.order {
+        SortOrder::Desc => {
+            |a, b: &liquid::value::Scalar| b.partial_cmp(&a).unwrap_or(Ordering::Equal)
+        }
+        SortOrder::Asc => {
+            |a: &liquid::value::Scalar, b| a.partial_cmp(&b).unwrap_or(Ordering::Equal)
+        }
+        SortOrder::None => {
+            // when built, order is set like this:
+            // `order.unwrap_or(SortOrder::Desc);` so it's unreachable
+            unreachable!("Sort order should have default value when constructing PaginationConfig")
+        }
+    };
+    let mut posts = posts.clone();
+    posts.sort_by(|a, b| {
+        let keys = &config.sort_by;
+        let mut cmp = Ordering::Less;
+        for k in keys {
+            cmp = match (extract_value(a, &k), extract_value(b, &k)) {
+                (Some(a), Some(b)) => order(a, b),
+                (None, None) => Ordering::Equal,
+                (_, None) => Ordering::Greater,
+                (None, _) => Ordering::Less,
+            };
+            if cmp != Ordering::Equal {
+                return cmp;
+            }
+        }
+        cmp
+    });
+    posts
+}
+
+fn create_all_paginators(
+    all_posts: &Vec<&liquid::value::Value>,
+    doc: &Document,
+    pagination_cfg: &PaginationConfig,
+) -> Result<Vec<Paginator>> {
+    let total_pages = all_posts.len();
+    // f32 used here in order to not lose information to ceil the result,
+    // otherwise we can lose an index
+    let total_indexes = (total_pages as f32 / pagination_cfg.per_page as f32).ceil() as usize;
+    let paginators: Result<Vec<_>> = (0..total_indexes)
```

Comment:
I think the idea is we could do something like
```rust
let paginators: Result<Vec<_>> = all_posts.chunks(pagination_cfg.per_page).enumerate().map(|(i, chunk)| {
            create_paginator(
                i,
                total_indexes,
                total_pages,
                &pagination_cfg,
                &doc,
                &chunk,
            )
});
```
So we are passing a `&[&Value]` to create_paginator that contains only the posts relevant for that paginator.